typeconverter.go

```diff
@@ -0,0 +1,125 @@
+package avro
+
+import (
+	"sync"
+)
+
+// TypeConverter represents a data type converter.
+type TypeConverter interface {
+	// Type returns the Avro data type of the type converter.
+	Type() Type
+
+	// LogicalType returns the Avro logical data type of the type converter.
+	LogicalType() LogicalType
+
+	// EncodeTypeConversion is the type conversion function called before encoding to Avro.
+	EncodeTypeConversion(in any) (any, error)
+
+	// DecodeTypeConversion is the type conversion function called after decoding from Avro.
+	DecodeTypeConversion(in any) (any, error)
+}
+
+type specificType struct {
+	typ Type
+	lt  LogicalType
+}
+
+// TypeConverters holds the user-provided type conversion functions.
+type TypeConverters struct {
+	convs sync.Map // map[specificType]TypeConverter
+}
+
+// NewTypeConverters creates a new type converter.
+func NewTypeConverters() *TypeConverters {
+	return &TypeConverters{}
+}
+
+// RegisterTypeConverters registers type converters for converting the data types during encoding and decoding.
+func (c *TypeConverters) RegisterTypeConverters(convs ...TypeConverter) {
+	for _, conv := range convs {
+		if typ := conv.Type(); len(typ) > 0 {
+			c.convs.Store(specificType{typ: conv.Type(), lt: conv.LogicalType()}, conv)
+		}
+	}
+}
+
+// EncodeConvert runs the encode type conversion function for the given value and schema.
+func (c *TypeConverters) EncodeConvert(in any, schema Schema) (any, error) {
+	typ := schema.Type()
+	lt := getLogicalType(schema)
+	conv, ok := c.convs.Load(specificType{typ: typ, lt: lt})
+	if !ok {
+		return in, nil
+	}
+
+	return conv.(TypeConverter).EncodeTypeConversion(in)
+}
+
+// DecodeConvert runs the decode type conversion function for the given schema and value.
+func (c *TypeConverters) DecodeConvert(schema Schema, in any) error {
+	pObj, ok := in.(*any)
+	if !ok {
+		return nil
+	}
+
+	typ := schema.Type()
+	lt := getLogicalType(schema)
+	conv, ok := c.convs.Load(specificType{typ: typ, lt: lt})
+	if !ok {
+		return nil
+	}
+
+	val, err := conv.(TypeConverter).DecodeTypeConversion(*pObj)
+	if err != nil {
+		*pObj = nil
+		return err
+	}
+
+	*pObj = val
+
+	return nil
+}
+
+// RegisterTypeConverters registers type converters for encoding and decoding the data type specified in the converter.
+func RegisterTypeConverters(convs ...TypeConverter) {
+	DefaultConfig.RegisterTypeConverters(convs...)
+}
+
+// TypeConversionFn is the function signature of the user-provided type conversion function.
+type TypeConversionFn func(in any) (any, error)
+
+// CustomTypeConverter is a helper struct to reduce boilerplate in user code.
+//
+// Implements the TypeConverter interface.
+type CustomTypeConverter struct {
```

Comment:
The name is not overly descriptive, perhaps `TypeConverterFuncs` is more to the point. Also not sure the `TypeConversionFn` type is needed.

Reply:
Changed and removed.